File: frontend/src/types/project.ts
/**
 * Project TypeScript Type Definitions
 *
 * WHAT: Type definitions for project management features.
 *
 * WHY: Ensures type safety for project CRUD operations,
 * API requests/responses, and UI components.
 *
 * HOW: Types match the backend Pydantic schemas for consistency.
 */

import type { OrgScopedEntity, ProjectStatus, ProjectPriority } from './common';

/**
 * Project entity
 *
 * WHAT: Full project data as returned by the API.
 *
 * WHY: Includes all fields for project display and management.
 */
export interface Project extends OrgScopedEntity {
  name: string;
  description: string | null;
  status: ProjectStatus;
  priority: ProjectPriority;
  estimated_hours: number | null;
  actual_hours: number | null;
  start_date: string | null;
  due_date: string | null;
  completed_at: string | null;
  is_active: boolean;
  is_overdue: boolean;
  hours_remaining: number | null;
}

/**
 * Project creation request
 *
 * WHAT: Data required to create a new project.
 *
 * WHY: Matches backend ProjectCreate schema.
 */
export interface ProjectCreateRequest {
  name: string;
  description?: string | null;
  priority?: ProjectPriority;
  estimated_hours?: number | null;
  start_date?: string | null;
  due_date?: string | null;
}

/**
 * Project update request
 *
 * WHAT: Data for updating an existing project.
 *
 * WHY: All fields optional for partial updates.
 */
export interface ProjectUpdateRequest {
  name?: string;
  description?: string | null;
  priority?: ProjectPriority;
  estimated_hours?: number | null;
  actual_hours?: number | null;
  start_date?: string | null;
  due_date?: string | null;
}

/**
 * Project status update request
 *
 * WHAT: Data for changing project status.
 *
 * WHY: Separate endpoint for status changes with audit logging.
 */
export interface ProjectStatusUpdateRequest {
  status: ProjectStatus;
}

/**
 * Project list response
 *
 * WHAT: Paginated list of projects.
 *
 * WHY: Backend returns skip/limit instead of page/pages.
 */
export interface ProjectListResponse {
  items: Project[];
  total: number;
  skip: number;
  limit: number;
}

/**
 * Project statistics
 *
 * WHAT: Aggregated project metrics.
 *
 * WHY: Dashboard widgets and summary views.
 */
export interface ProjectStats {
  total: number;
  active: number;
  by_status: Record<string, number>;
  overdue: number;
}

/**
 * Project list query parameters
 *
 * WHAT: Filters and pagination for project list.
 *
 * WHY: Type-safe query building.
 */
export interface ProjectListParams {
  skip?: number;
  limit?: number;
  status?: ProjectStatus;
  priority?: ProjectPriority;
  active_only?: boolean;
  search?: string;
}

/**
 * Status display configuration
 *
 * WHAT: UI metadata for project statuses.
 *
 * WHY: Consistent status badges and colors.
 */
export const PROJECT_STATUS_CONFIG: Record<
  ProjectStatus,
  { label: string; color: string; bgColor: string }
> = {
  draft: { label: 'Draft', color: 'text-gray-600', bgColor: 'bg-gray-100' },
  proposal_sent: { label: 'Proposal Sent', color: 'text-blue-600', bgColor: 'bg-blue-100' },
  approved: { label: 'Approved', color: 'text-green-600', bgColor: 'bg-green-100' },
  in_progress: { label: 'In Progress', color: 'text-yellow-600', bgColor: 'bg-yellow-100' },
  on_hold: { label: 'On Hold', color: 'text-orange-600', bgColor: 'bg-orange-100' },
  completed: { label: 'Completed', color: 'text-emerald-600', bgColor: 'bg-emerald-100' },
  cancelled: { label: 'Cancelled', color: 'text-red-600', bgColor: 'bg-red-100' },
};

/**
 * Priority display configuration
 *
 * WHAT: UI metadata for project priorities.
 *
 * WHY: Consistent priority badges and colors.
 */
export const PROJECT_PRIORITY_CONFIG: Record<
  ProjectPriority,
  { label: string; color: string; bgColor: string }
> = {
  low: { label: 'Low', color: 'text-gray-600', bgColor: 'bg-gray-100' },
  medium: { label: 'Medium', color: 'text-blue-600', bgColor: 'bg-blue-100' },
  high: { label: 'High', color: 'text-orange-600', bgColor: 'bg-orange-100' },
  urgent: { label: 'Urgent', color: 'text-red-600', bgColor: 'bg-red-100' },
};

File: backend/app/schemas/project.py
"""
Pydantic schemas for project endpoints.

WHAT: Request/response schemas for project management API.

WHY: Schemas define API contracts for project operations:
1. Validate incoming request data
2. Document API for OpenAPI/Swagger
3. Provide type safety for handlers
4. Control which fields are exposed in responses

HOW: Uses Pydantic v2 with Field validators and ORM mode for SQLAlchemy.
"""

from datetime import datetime
from enum import Enum
from typing import Optional
from pydantic import BaseModel, Field, field_validator


class ProjectStatus(str, Enum):
    """
    Project lifecycle status.

    WHY: Mirrors the SQLAlchemy enum for API consistency.
    """

    DRAFT = "draft"
    PROPOSAL_SENT = "proposal_sent"
    APPROVED = "approved"
    IN_PROGRESS = "in_progress"
    ON_HOLD = "on_hold"
    COMPLETED = "completed"
    CANCELLED = "cancelled"


class ProjectPriority(str, Enum):
    """
    Project priority level.

    WHY: Mirrors the SQLAlchemy enum for API consistency.
    """

    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    URGENT = "urgent"


class ProjectCreate(BaseModel):
    """
    Project creation request schema.

    WHAT: Validates data for creating a new project.

    WHY: Ensures all required fields are present and valid before
    database insertion. Projects start in DRAFT status automatically.
    """

    name: str = Field(
        ...,
        min_length=1,
        max_length=255,
        description="Project name/title",
    )
    description: str | None = Field(
        default=None,
        max_length=5000,
        description="Detailed project description",
    )
    priority: ProjectPriority = Field(
        default=ProjectPriority.MEDIUM,
        description="Project priority level",
    )
    estimated_hours: float | None = Field(
        default=None,
        ge=0,
        description="Estimated hours for project completion",
    )
    start_date: datetime | None = Field(
        default=None,
        description="Planned start date (ISO 8601 format)",
    )
    due_date: datetime | None = Field(
        default=None,
        description="Target completion date (ISO 8601 format)",
    )

    @field_validator("due_date")
    @classmethod
    def due_date_after_start_date(cls, v: datetime | None, info) -> datetime | None:
        """
        Validate due_date is after start_date.

        WHY: Logical constraint - can't have due date before start date.
        """
        if v is not None and info.data.get("start_date") is not None:
            if v < info.data["start_date"]:
                raise ValueError("due_date must be after start_date")
        return v

    class Config:
        json_schema_extra = {
            "example": {
                "name": "Website Automation Project",
                "description": "Automate lead capture and CRM integration",
                "priority": "high",
                "estimated_hours": 40,
                "start_date": "2025-01-15T00:00:00",
                "due_date": "2025-02-15T00:00:00",
            }
        }


class ProjectUpdate(BaseModel):
    """
    Project update request schema.

    WHAT: Validates data for updating an existing project.

    WHY: Allows partial updates - only provided fields are modified.
    Status changes use a separate endpoint for workflow control.
    """

    name: str | None = Field(
        default=None,
        min_length=1,
        max_length=255,
        description="Project name/title",
    )
    description: str | None = Field(
        default=None,
        max_length=5000,
        description="Detailed project description",
    )
    priority: ProjectPriority | None = Field(
        default=None,
        description="Project priority level",
    )
    estimated_hours: float | None = Field(
        default=None,
        ge=0,
        description="Estimated hours for project completion",
    )
    actual_hours: float | None = Field(
        default=None,
        ge=0,
        description="Actual hours spent on project",
    )
    start_date: datetime | None = Field(
        default=None,
        description="Planned start date (ISO 8601 format)",
    )
    due_date: datetime | None = Field(
        default=None,
        description="Target completion date (ISO 8601 format)",
    )

    class Config:
        json_schema_extra = {
            "example": {
                "name": "Updated Project Name",
                "priority": "urgent",
                "actual_hours": 25,
            }
        }


class ProjectStatusUpdate(BaseModel):
    """
    Project status update request schema.

    WHAT: Validates status change requests.

    WHY: Status changes have business implications:
    - COMPLETED sets completed_at timestamp
    - Certain transitions may trigger notifications
    - Separate endpoint allows audit logging of status changes
    """

    status: ProjectStatus = Field(
        ...,
        description="New project status",
    )

    class Config:
        json_schema_extra = {
            "example": {
                "status": "in_progress",
            }
        }


class ProjectResponse(BaseModel):
    """
    Project response schema.

    WHAT: Structure for project data in API responses.

    WHY: Controls which fields are exposed and adds computed properties.
    Uses ORM mode for automatic SQLAlchemy model conversion.
    """

    id: int = Field(..., description="Project ID")
    name: str = Field(..., description="Project name/title")
    description: str | None = Field(None, description="Project description")
    status: ProjectStatus = Field(..., description="Current project status")
    priority: ProjectPriority = Field(..., description="Project priority level")
    org_id: int = Field(..., description="Organization ID")
    estimated_hours: float | None = Field(None, description="Estimated hours")
    actual_hours: float | None = Field(None, description="Actual hours spent")
    start_date: datetime | None = Field(None, description="Planned start date")
    due_date: datetime | None = Field(None, description="Target completion date")
    completed_at: datetime | None = Field(None, description="Completion timestamp")
    created_at: datetime = Field(..., description="Creation timestamp")
    updated_at: datetime = Field(..., description="Last update timestamp")
    is_active: bool = Field(..., description="Whether project is active")
    is_overdue: bool = Field(..., description="Whether project is past due date")
    hours_remaining: float | None = Field(None, description="Estimated hours remaining")

    class Config:
        from_attributes = True
        json_schema_extra = {
            "example": {
                "id": 1,
                "name": "Website Automation Project",
                "description": "Automate lead capture and CRM integration",
                "status": "in_progress",
                "priority": "high",
                "org_id": 1,
                "estimated_hours": 40,
                "actual_hours": 25,
                "start_date": "2025-01-15T00:00:00",
                "due_date": "2025-02-15T00:00:00",
                "completed_at": None,
                "created_at": "2025-01-10T10:30:00",
                "updated_at": "2025-01-20T15:45:00",
                "is_active": True,
                "is_overdue": False,
                "hours_remaining": 15,
            }
        }


class ProjectListResponse(BaseModel):
    """
    Paginated project list response schema.

    WHAT: Wrapper for paginated project list responses.

    WHY: Provides pagination metadata alongside items for:
    - Client-side pagination UI
    - Total count for progress indicators
    - Consistent list response format across API
    """

    items: list[ProjectResponse] = Field(..., description="List of projects")
    total: int = Field(..., description="Total number of projects matching filters")
    skip: int = Field(..., description="Number of items skipped (offset)")
    limit: int = Field(..., description="Maximum items per page")

    class Config:
        json_schema_extra = {
            "example": {
                "items": [
                    {
                        "id": 1,
                        "name": "Website Automation",
                        "status": "in_progress",
                        "priority": "high",
                        "org_id": 1,
                        "is_active": True,
                        "is_overdue": False,
                    }
                ],
                "total": 15,
                "skip": 0,
                "limit": 10,
            }
        }


class ProjectStats(BaseModel):
    """
    Project statistics response schema.

    WHAT: Aggregated project metrics for dashboards.

    WHY: Provides quick overview of project states without
    fetching all project data. Useful for dashboard widgets.
    """

    total: int = Field(..., description="Total projects")
    active: int = Field(..., description="Active (not completed/cancelled) projects")
    by_status: dict[str, int] = Field(
        ..., description="Count of projects by status"
    )
    overdue: int = Field(..., description="Number of overdue projects")

    class Config:
        json_schema_extra = {
            "example": {
                "total": 25,
                "active": 18,
                "by_status": {
                    "draft": 3,
                    "proposal_sent": 2,
                    "approved": 4,
                    "in_progress": 8,
                    "on_hold": 1,
                    "completed": 5,
                    "cancelled": 2,
                },
                "overdue": 2,
            }
        }
